### DISCUSSION TITLE
Working hours plugin vs core / support database migrations for plugins

### DISCUSSION THREAD

**Discussion by freakimkaefig** [kimai/kimai#1854] [General]
**Is your feature request related to a problem? Please describe.**
I implemented a way to track daily/weekly/monthly/yearly working hours including exceptions (vacation, holiday) in a fork i'm using ([https://github.com/freakimkaefig/kimai2](https://github.com/freakimkaefig/kimai2)) as i did not find any solution to create custom database tables from plugins.  
In order to handle the exceptions i added an additional database table.

**Describe the solution you'd like**
I would like to provide my implementation as a plugin. Is there a possibility to create database tables for plugins? I couldn't find any in the [docs](https://www.kimai.org/documentation/plugins.html).

**Describe alternatives you've considered**
Maybe i can send you a pull request and you integrate the feature into the kimai core ;-)

**Comment by kevinpapst** [comment]
Its not a big deal to use migrations from a plugin. 
I always build a command to simplify the installation process.

Please have a look at the DemoBundle: https://github.com/Keleo/DemoBundle

And especially the last commit which I added right now to demonstrate the usage of installation command in combination with a doctrine migration. That should cover everything you need: 
https://github.com/Keleo/DemoBundle/commit/da131db6068202c692566c87f713e56708c650e5

Really looking forward to see your plugin in the Marketplace 😄 

Let me know if you need more help and don't hesitate to open new issues if you need help.

**Comment by kevinpapst** [comment]
Your change is quite large, I need some time to go through it to understand it properly.
From a quick glance it looks very promising 👍 

Regarding the working hours: I planned on having such a setting/integration as well (obviously required in the long run for a time-tracker). 

I wanted to have a weekly target hour setting and a start and end date for that, because working contracts can change over time and we need to be able to calculate that for the past as well.
So storing them as user preference wouldn't work in the long run. If the services are proper planned we can still exchange the use of user preference with a dedicated database in a later version.

Not sure though that everything you are doing there is possible as plugin and a feature like that was planned for the core as well, so it might be better to integrate it as PR to avoid duplication. I have quite some remarks and change requests... depends on you if you want to go that way together with me 😁 

So give me some time to have a chance on testing it and until then, please let me know what you think about the future of your change.

**Comment by freakimkaefig** [comment]
Thanks for the quick reply. I will have a look into the DemoBundle plugin code for migrations and will try to integrate my feature as a plugin. So i can get a better understanding for the code base 😉
I'm actually quite new to Symfony etc.

In the long run it would be cool working on a core functionality to handle working hours/overtime. Feel free to send me some ideas or concepts you are planning. I can try to update the code to fit. The idea of storing weekly working hours for a period sounds great and for the sake of flexibility way better than my approach using preferences 😁

**Comment by therges** [comment]
I was looking for that!
10 € each for you - if that is implemented!
